clang/lib/DPCT/ASTTraversal.cpp

```diff
@@ -11894,11 +11897,62 @@ void CooperativeGroupsFunctionRule::registerMatcher(MatchFinder &MF) {
                                                hasAttr(attr::CUDAGlobal))))))
           .bind("FuncCall"),
       this);
+  MF.addMatcher(
+      declRefExpr(
+          hasAncestor(
+              implicitCastExpr(
+                  hasImplicitDestinationType(qualType(hasCanonicalType(
+                      recordType(hasDeclaration(cxxRecordDecl(hasName(
+                          "cooperative_groups::__v1::thread_group"))))))))))
+          .bind("declRef"),
+      this);
 }
 
 void CooperativeGroupsFunctionRule::runRule(
     const MatchFinder::MatchResult &Result) {
   const CallExpr *CE = getNodeAsType<CallExpr>(Result, "FuncCall");
+  const DeclRefExpr *DR = getNodeAsType<DeclRefExpr>(Result, "declRef");
+  const SourceManager &SM = DpctGlobalInfo::getSourceManager();
+  if (DR) {
+    std::string DREName = DR->getType().getCanonicalType().getAsString();
+    std::string ReplacedStr = "";
+    if (DpctGlobalInfo::getAssumedNDRangeDim() == 1) {
+      auto FD = DpctGlobalInfo::getParentFunction(DR);
+      if (!FD)
+        return;
+      auto DFI = DeviceFunctionDecl::LinkRedecls(FD);
+      auto Index = DpctGlobalInfo::getCudaKernelDimDFIIndexThenInc();
+      ReplacedStr = "<{{NEEDREPLACEG" + std::to_string(Index) + "}}>";
+      DpctGlobalInfo::insertCudaKernelDimDFIMap(Index, DFI);
+    } else {
+      ReplacedStr = "<3>";
+    }
+
+    if (DREName.find("cooperative_groups::__v1::thread_block_tile<32>") !=
+        std::string::npos) {
+      ReplacedStr = "sycl::sub_group";
+    } else if (DREName == "class cooperative_groups::__v1::thread_block") {
+      ReplacedStr = "sycl::group" + ReplacedStr;
+    } else {
+      ReplacedStr = "dpct::experimental::logical_group" + ReplacedStr;
```

Comment:
For dpct::experimental namespace hardcoded, do we need an option to control it, like option "--use-experimental-features=<value>" does?

Reply:
I think here we can use `DpctGlobalInfo::useLogicalGroup()`

Reply:
Updated